services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,514 @@
+     *         The <em>status</em> will be
+     *         <ul>
+     *         <li><em>204 No Content</em> if the tenant has been updated successfully.</li>
+     *         <li><em>403 forbidden</em> if tenant modification is forbidden due to configuration settings</li>
+     *         <li><em>404 Not Found</em> if no tenant with the given identifier and version exists.</li>
+     *         <li><em>409 Conflict</em> if provided trusted certificate authority is already present in another
+     *         tenant.</li>
+     *         <li><em>412 Precondition Failed</em> if the version exists but mismatch the current tenants version.</li>
+     *         </ul>
+     */
+    @Override
+    public Future<OperationResult<Void>> updateTenant(final String tenantId, final Tenant tenantObj,
+            final Optional<String> resourceVersion, final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(tenantObj);
+        Objects.requireNonNull(resourceVersion);
+        Objects.requireNonNull(span);
+
+        return MongoDbDeviceRegistryUtils.isAllowedToModify(config, tenantId)
+                .compose(ok -> findTenant(tenantId))
+                .compose(tenantDto -> checkConflictingCATenants(Optional.of(tenantDto),
+                        tenantObj.getTrustedCertificateAuthoritySubjectDNs(), span))
+                .compose(tenantDto -> processUpdateTenant(tenantId, tenantObj, resourceVersion, span))
+                .recover(error -> Future.succeededFuture(MongoDbDeviceRegistryUtils.mapErrorToResult(error, span)));
+    }
+
+    private Future<OperationResult<Void>> processUpdateTenant(final String tenantId, final Tenant newTenant,
+            final Optional<String> resourceVersion, final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(newTenant);
+        Objects.requireNonNull(resourceVersion);
+        Objects.requireNonNull(span);
+
+        final JsonObject updateTenantQuery = resourceVersion
+                .map(version -> new MongoDbDocumentBuilder().withVersion(version))
+                .orElse(new MongoDbDocumentBuilder())
+                .withTenantId(tenantId)
+                .document();
+
+        final Promise<JsonObject> updateTenantPromise = Promise.promise();
+        final TenantDto newTenantDto = new TenantDto(tenantId, newTenant,
+                new Versioned<>(newTenant).getVersion());
+        mongoClient.findOneAndReplaceWithOptions(config.getCollectionName(), updateTenantQuery,
+                JsonObject.mapFrom(newTenantDto), new FindOptions(), new UpdateOptions().setReturningNewDocument(true),
+                updateTenantPromise);
+
+        return updateTenantPromise.future()
+                .compose(updateResult -> Optional.ofNullable(updateResult)
+                        .map(updated -> {
+                            span.log(String.format("successfully updated device [%s]", tenantId));
```

Comment:
Please change it to `successfully updated tenant` not `device`.

Reply:
fixed